### DISCUSSION TITLE
FluidSynth as stereo DrumKit

### DISCUSSION THREAD

**Discussion by dbadb** [FluidSynth/fluidsynth#1108] [General]
Greetings, I'm trying to figure out how/whether I can get use FluidSynth to simulate a stereo drumkit where each "note" of a percussion preset can reside in a unique stereo position.  Any pointers are much appreciated!

Here's what I've learned so far:

- the default channel setup described by GM puts the percussion sound bank on MIDI channel 9.
- the MIDI pan request works in Fluid (yay). If I precede any note (programmatically) with a CC 10, I can place the sound in stereo field.

**But**

- if I have two "notes" (say Kick (note 36) and Snare (note 40) that play _simultaneously_, they both appear to "inherit" the current pan-setting and this destroys the desired effect.

Potential Solution

- since the pan message is multi-channel I'd hoped to assign different notes to different MIDI channels.  Now, whether by static or dynamic assignment, I could cause the  Kick and Snare to keep their correct pan position whether hit simultaneously or not.  This solution requires that I be able to assign the special drum banks to any or all MIDI channels.  And this is where I'm currently blocked.
I have discovered the different modes for synth.midi-bank-select=mma and have used this to successfully change the behavior of channel 0.  Now I have two drumkit channels (0 and 9), each with a unique stereo location.   If I now try for three drumkit channels by sending CC messages (0 and 32), I neither see an error message nor hear my drums on channel 1 (I get the default piano value).

In pseudo-code:
```
// FluidR3_GM.sf2, midi-bank-select=mma
cc(0, bank/128, chan);
cc(32, bank%128, chan);
progChange(chan, prog); /* 0 is Standard, 8: Room, 25: TR808 */
runOfNotes(chan, 5); 
```

I've seen some indication that GM supports up-to two percussion channels.  I wonder if that's what's at issue here?

Any pointers are much appreciated!

**Comment by derselbst** [comment]
Yes, the SoundFont spec allows you to do this. fluidsynth implements the SoundFont spec, ergo fluidsynth can do it.

However, you need to put off your programmer glasses and put on your SoundFont designer glasses: In the SoundFont(s) you want to use, you must specify the initial Pan position of each and every instrument zone (i.e. sample) of your drum instrument(s).

Coming back to your Kick / Snare example, you would open the drum-instrument containing those two sample by using your favorite SoundFont editor. You would select the zone which contains the Kick sample and set its Pan to -50. Similarly, you would select the zone which contains the Snare and set its Pan to +50.

![image](https://user-images.githubusercontent.com/8152480/171952230-bacdba6a-d7ba-4433-87e4-737486c6bb13.png)


Doing it on a SoundFont level is some amount of work, yes. But this solution is portable among SoundFont compatible synths.

When going the conventional programming way by sending CCs, you'll run into the error you've described.

If you insist on a programmatical solution, you might want to have a look at MIDI Polyphonic Expression. It allows you to control notes indiviually (because it dynamically assigns them to their own MIDI channel). However, MPE is currently not implemented in fluidsynth and I don't know about details.

> I have discovered the different modes for synth.midi-bank-select=mma and have used this to successfully change the behavior of channel 0.

Pls. do not confuse the channel mode with the bank selection behavior. Even though `synth.midi-bank-select=mma` allows you to explicitly select the drum bank and preset, it is still treated as melodic channel, unless you apply `fluid_synth_set_channel_type()` accordingly. The difference between drum and melodic channels is the preset fallback mechanism used when no preset for a given MIDI bank + program could be found... which could be a possible explanation why you don't hear drums on channel 1.

**Reply by dbadb** [reply]
Tom M - thank you very much for this very-useful information.

In my case, the goal is to employ a range of programs in a variety of sound fonts to a range of circumstances which makes the edit-soundfile-approach untenable. For example, I'd like to select the Jazz drumkit for one tune, then the Standard drumkit for another.  And for Jazz, the kick is on the left and for Standard the kick is on the right.

Your point about `fluid_synth_set_channeL_type()` seems very relevant/important and should be viable for my setup.    I just
did a quick test and it appears to have solved my problem so far.

Thanks *very much* for pointing me in the right direction!

regards, Dana B
